monitoring/queuemon_test.go

```diff
@@ -0,0 +1,186 @@
+	return queue.BufferConfig{}
+}
+
+// Producer doesn't do anything
+func (tq TestMetricsQueue) Producer(_ queue.ProducerConfig) queue.Producer {
+	return nil
+}
+
+// Consumer doesn't do anything
+func (tq TestMetricsQueue) Consumer() queue.Consumer {
+	return nil
+}
+
+// Close Doesn't do anything
+func (tq TestMetricsQueue) Close() error {
+	return nil
+}
+
+// Metrics spoofs the metrics output
+func (tq *TestMetricsQueue) Metrics() (queue.Metrics, error) {
+	//fmt.Printf("Got queue state: %#v\n", tq)
+	tq.metricState.EventCount = opt.UintWith(tq.metricState.EventCount.ValueOr(0) + 1)
+
+	if tq.metricState.EventCount.ValueOr(0) > tq.limit {
+		tq.metricState.EventCount = opt.UintWith(0)
+	}
+
+	return tq.metricState, nil
+}
+
+// ============ mocked output reporter
+
+// simple wrapper to return a generic config object
+func initMonWithconfig(interval int, name string) Config {
+	return Config{
+		Interval: time.Millisecond * time.Duration(interval),
+		Enabled:  true,
+		ExpvarOutput: expvar.Config{
+			Enabled: true,
+			Port:    8081,
+			Host:    "",
+			Name:    name,
+		},
+	}
+
+}
+
+// fetch the expvar data from the http endpoint and return the final queue object to test the metrics outputs
+func fetchExpVars(t *testing.T, client http.Client, endpoint string) expvarQueue {
+	resp, err := client.Get(endpoint) //nolint:noctx //this is a test, with a timeout
+	assert.NoError(t, err)
+	defer resp.Body.Close()
+	assert.Equal(t, resp.StatusCode, 200, "expvar endpoint did not return 200: %#v", resp)
+	httpResp, err := io.ReadAll(resp.Body)
+	assert.NoError(t, err)
+	raw := struct {
+		Memstats runtime.MemStats
+		Cmdline  []string
+		Queue    expvarQueue
+	}{}
+	err = json.Unmarshal(httpResp, &raw)
+
+	assert.NoError(t, err)
+
+	return raw.Queue
+}
+
+// actual tests
+
+func TestSetupMonitor(t *testing.T) {
+	monitor := initMonWithconfig(1, "test")
+	queue := NewTestQueue(10)
+	mon, err := NewFromConfig(monitor, queue)
+	assert.NoError(t, err)
+	mon.Watch()
+	mon.End()
+
+}
+
+func TestReportedEvents(t *testing.T) {
+	monitor := initMonWithconfig(1, "queue")
+
+	var maxEvents uint64 = 10
+	queue := NewTestQueue(maxEvents)
+	mon, err := NewFromConfig(monitor, queue)
+	assert.NoError(t, err)
+	mon.Watch()
+
+	var limitCount int
+	var gotQueueIsFull bool
+	var queueFullCount int
+
+	t.Logf("listening for events...")
+	// once we have maxEvents, we can properly
+	// use the expvar endpoint to check the output
+	client := http.Client{
+		Timeout: 5 * time.Second,
+	}
+	endpoint := "http://localhost:8081/debug/vars"
+
+	// Sit and wait until we have interesting data we can test
+	for {
+		result := fetchExpVars(t, client, endpoint)
+		t.Logf("Got raw result: %#v", result)
+		if result.IsFull {
+			gotQueueIsFull = true
+			queueFullCount = result.CurrentLevel
+			limitCount = result.LimitCount
+			break
+		}
+
+	}
+	mon.End()
+	assert.True(t, gotQueueIsFull, "Did not report a full queue")
```

Comment:
How could that not be true if when you exit the loop, you set it to true?
This test is only testing the other 2 values, `limitCount` and `queueFullCount`.